src/ranking.jl

```diff
@@ -6,50 +6,63 @@
 # The implementations here follow this wikipedia page.
 #
 
-
 function _check_randparams(rks, x, p)
     n = length(rks)
     length(x) == length(p) == n || raise_dimerror()
     return n
 end
 
+# ranking helper function: calls sortperm(x) and then ranking method f!
+function _rank(f!, x::AbstractArray, R::Type=Int; sortkwargs...)
+    rks = similar(x, R)
+    ord = reshape(sortperm(vec(x); sortkwargs...), size(x))
+    return f!(rks, x, ord)
+end
 
+# ranking helper function for arrays with missing values
+function _rank(f!, x::AbstractArray{>: Missing}, R::Type=Int; sortkwargs...)
+    inds = findall(!ismissing, vec(x))
+    isempty(inds) && return missings(R, size(x))
+    T = nonmissingtype(eltype(x))
+    xv = Vector{T}(undef, length(inds))
+    @inbounds for (i, ind) in enumerate(inds)
+        xv[i] = x[ind]
+    end
+    rks = missings(R, size(x))
+    ordv = sortperm(xv; sortkwargs...)
+    f!(view(rks, inds), xv, ordv)
+    return rks
+end
 
 # Ordinal ranking ("1234 ranking") -- use the literal order resulted from sort
-function ordinalrank!(rks::AbstractArray, x::AbstractArray, p::IntegerArray)
-    n = _check_randparams(rks, x, p)
-
-    if n > 0
-        i = 1
-        while i <= n
-            rks[p[i]] = i
-            i += 1
-        end
+function _ordinalrank!(rks::AbstractArray, x::AbstractArray, p::IntegerArray)
+    _check_randparams(rks, x, p)
+    @inbounds for i in eachindex(p)
+        rks[p[i]] = i
     end
-
     return rks
 end
 
 
 """
-    ordinalrank(x; lt = isless, rev::Bool = false)
+    ordinalrank(x; sortkwargs...)
 
 Return the [ordinal ranking](https://en.wikipedia.org/wiki/Ranking#Ordinal_ranking_.28.221234.22_ranking.29)
 ("1234" ranking) of an array. The `lt` keyword allows providing a custom "less
 than" function; use `rev=true` to reverse the sorting order.
 All items in `x` are given distinct, successive ranks based on their
-position in `sort(x; lt = lt, rev = rev)`.
+position in `sort(x; sortkwargs...)`.
 Missing values are assigned rank `missing`.
 """
-ordinalrank(x::AbstractArray; lt = isless, rev::Bool = false) =
-    ordinalrank!(Array{Int}(undef, size(x)), x, sortperm(x; lt = lt, rev = rev))
+ordinalrank(x::AbstractArray; sortkwargs...) =
+    _rank(_ordinalrank!, x; sortkwargs...)
 
 
 # Competition ranking ("1224" ranking) -- resolve tied ranks using min
-function competerank!(rks::AbstractArray, x::AbstractArray, p::IntegerArray)
+function _competerank!(rks::AbstractArray, x::AbstractArray, p::IntegerArray)
     n = _check_randparams(rks, x, p)
 
-    if n > 0
+    @inbounds if n > 0
```

Comment:
While you're at it, maybe replace these `while` loops with `for i in 2:n` and so on? They are hard to read currently.